modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/validators.js

```diff
@@ -0,0 +1,366 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+const ERROR_OBJECT_OF_TYPE = 'Expected object of one type.';
+const ERROR_ONE_OF = 'Expected one of the following values:';
+const ERROR_ONE_OF_TYPE = 'Expected one of given types.';
+
+function _getFunctionName(fn) {
+	if (!fn.name) {
+		const stringifiedName = fn.toString();
+
+		fn.name = stringifiedName.substring(9, stringifiedName.indexOf('('));
+	}
+
+	return fn.name;
+}
+
+function _isDefAndNotNull(value) {
+	return value !== undefined && value !== null;
+}
+
+/**
+ * Provides access to various type validators that will return an
+ * instance of Error when validation fails. Note that all type validators
+ * will also accept null or undefined values. To not accept these you should
+ * instead make your state property required.
+ */
+const validators = {
+	any: () => () => true,
+	array: buildTypeValidator('array'),
+
+	/**
+	 * Creates a validator that checks that the value it receives is an array
+	 * of items, and that all of the items pass the given validator.
+	 * @param {!Object} validator Validator to check each item against.
+	 * @return {!function()}
+	 */
+	arrayOf(validator) {
+		if (isInvalid(validators.object(validator))) {
+			throwConfigError('object', validator, 'arrayOf');
+		}
+
+		return maybe((value, name, context) => {
+			const result = validators.array(value, name, context);
```

Comment:
Could sweep through here inserting blank lines here where Brian might ask for them. This is auto-enforced on the Java side, I believe, but we haven't done it for JS yet (we have an open [issue for doing that](https://github.com/liferay/liferay-frontend-projects/issues/15)). In Brian's own words:

> In short, if you declare a var, add a line break before using it. This shows order of requirements better.

Another way to express the idea is "if lines at the same level of indentation can be in any order, they should be together with no intervening blanks; otherwise they should be separated".